// websocket-game-lobby/dist/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServerEvents = exports.DataStoreEvents = void 0;
/** Types to be used with event listeners attached to {@link DataStore} objects. */
var DataStoreEvents;
(function (DataStoreEvents) {
    /** Event fired when game is created. */
    DataStoreEvents["createGame"] = "createGame";
    /** Event fired when player or spectator leaves game. */
    DataStoreEvents["leaveGame"] = "leaveGame";
    /** Event fired when game is started. */
    DataStoreEvents["startGame"] = "startGame";
    /** Event fired when player is created. */
    DataStoreEvents["createPlayer"] = "createPlayer";
    /** Event fired when spectator is created. */
    DataStoreEvents["createSpectator"] = "createSpectator";
    /** Event fired when turn is created. */
    DataStoreEvents["createTurn"] = "createTurn";
    /** Event fired when current turn is ended. */
    DataStoreEvents["endCurrentTurn"] = "endCurrentTurn";
})(DataStoreEvents = exports.DataStoreEvents || (exports.DataStoreEvents = {}));
/** Types to be used with event listeners attached to the {@link WebSocketGameLobbyServer}. */
var ServerEvents;
(function (ServerEvents) {
    /** Event fired when game is created. */
    ServerEvents["create"] = "create";
    /** Event fired when game is joined. */
    ServerEvents["join"] = "join";
    /** Event fired when game is left. */
    ServerEvents["leave"] = "leave";
    /** Event fired when game is started. */
    ServerEvents["start"] = "start";
    /** Event fired when game is ended. */
    ServerEvents["end"] = "end";
})(ServerEvents = exports.ServerEvents || (exports.ServerEvents = {}));
